3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,342 @@
+        .failure
+    }
+}
+
+/**
+ * The PiiPseudonymizerEnrichment runs after all other enrichments to find fields that are configured as PII (personally
+ * identifiable information) and apply some anonymization (currently only psudonymizantion) on them. Currently a single
+ * strategy for all the fields is supported due to the config format, and there is only one implemented strategy,
+ * however the enrichment supports a strategy per field.
+ *
+ * The user may specify two types of fields POJO or JSON. A POJO field is effectively a scalar field in the
+ * EnrichedEvent, whereas a JSON is a "context" formatted field and it can be wither a scalar in the case of
+ * unstruct_event or an array in the case of derived_events and contexts
+ *
+ * @param fieldList a list of configured PiiFields
+ */
+case class PiiPseudonymizerEnrichment(fieldList: List[PiiField]) extends Enrichment {
+  def transformer(event: EnrichedEvent): Unit = fieldList.foreach(_.transform(event))
+}
+
+/**
+ * Specifies a field in POJO and the strategy that should be applied to it.
+ * @param strategy the strategy that should be applied
+ * @param fieldMutator the field mutator where the strategy will be applied
+ */
+final case class PiiPojo(strategy: PiiStrategy, fieldMutator: PiiConstants.Mutator) extends PiiField {
+  override def applyStrategy(fieldValue: String): String = if (fieldValue != null) strategy.scramble(fieldValue) else null
+}
+
+/**
+ * Specifies a strategy to use, a field mutator where the JSON can be found, a schema criterion to
+ * discriminate which contexts to apply this strategy to, and a json path within the contexts where this strategy will
+ * be applied (the path may correspond to multiple fields).
+ *
+ * @param strategy the strategy that should be applied
+ * @param fieldMutator the field mutator for the json field
+ * @param schemaCriterion the schema for which the strategy will be applied
+ * @param jsonPath the path where the strategy will be applied
+ */
+final case class PiiJson(strategy: PiiStrategy, fieldMutator: PiiConstants.Mutator, schemaCriterion: SchemaCriterion, jsonPath: String)
+    extends PiiField {
+  implicit val json4sFormats = DefaultFormats
+
+  override def applyStrategy(fieldValue: String): String =
+    if (fieldValue != null)
+      compact(render(parse(fieldValue).transformField {
+        case JField("data", JArray(contexts)) =>
+          ("data", JArray(contexts.map {
+            case JObject(context) => modifyObjectIfSchemaMatches(context)
+            case x                => x
+          }))
+        case JField("data", JObject(unstructEvent)) =>
+          ("data", modifyObjectIfSchemaMatches(unstructEvent))
+      }))
+    else null
+
+  private def modifyObjectIfSchemaMatches(context: List[(String, json4s.JValue)]): JObject = {
+    val fields: Map[String, JValue] = List("schema", "data").flatMap(k => context.toMap.get(k).map(v => (k -> v))).toMap
```

Comment:
I have trouble understanding why you have to go through all this, isn't the above equivalent to:

```scala
context.toMap.filterKeys(List("schema", "data").contains)
```

Reply:
good point

Reply:
it would be even better to extract out the List or do s => s == "schema" || s == "data", that would avoid having to create n lists.

Reply:
👍